### PR TITLE
Added rename_coords and rename_dims

### PR DESCRIPTION
**Pull request by jukent**
- [ ] Closes #xxxx
 - [ ] Tests added
 - [ ] Fully documented, including `whats-new.rst` for all changes and `api.rst` for new API

### CODE REVIEW

File: xarray/core/dataset.py [hunk 1]
```diff
@@ -2285,10 +2287,88 @@ def rename(self, name_dict=None, inplace=None, **names):
...
+            if k not in self.dims:
+                raise ValueError("cannot rename %r because it is not a "
+                                 "dimension in this dataset" % k)
+
+        dims = self._rename_dims(name_dict)
+        indexes = self._rename_indexes(name_dict)
+        return self._replace(dims=dims, indexes=indexes, inplace=inplace)
+
+
+    def rename_coords(self, name_dict=None, inplace=None, **names):
```
Comment (shoyer) [inline xarray/core/dataset.py]: I think it probably makes sense to call this `rename_vars()` (though we could also define a method for renaming coordinates only)

Comment (jukent) [reply xarray/core/dataset.py]: I was wondering about that. rename_vars seemed to be more consistent with the code.

**Comment by jukent** [comment]
pydata#3026

There is some unexpected behavior. The new rename_dims function does not change the dimensions of the variables. Working on this now.

**Comment by jukent** [comment]
> #3026
> 
> There is some unexpected behavior. The new rename_dims function does not change the dimensions of the variables. Working on this now.

Similarly rename_vars will rename the variable dimensions and variables, but not the dataset dimensions. I thought I was dealing with this with the _rename_indexes. But will have to keep working.

File: xarray/core/dataset.py [hunk 2]
```diff
@@ -2285,10 +2287,88 @@ def rename(self, name_dict=None, inplace=None, **names):
...
+            if k not in self.dims:
+                raise ValueError("cannot rename %r because it is not a "
+                                 "dimension in this dataset" % k)
+
+        dims = self._rename_dims(name_dict)
+        indexes = self._rename_indexes(name_dict)
+        return self._replace(dims=dims, indexes=indexes, inplace=inplace)
+
+
+    def rename_vars(self, name_dict=None, inplace=None, **names):
```
Comment (max-sixty) [inline xarray/core/dataset.py]: What do people think about `vars` vs `variables` here?

I'm +0.2 on `variables` for consistency with `merge_variables` etc (but only did a quick search, I would update if there were more public `vars` methods around)

File: xarray/core/dataset.py [hunk 3]
```diff
@@ -2285,10 +2287,88 @@ def rename(self, name_dict=None, inplace=None, **names):
...
+        """Returns a new object with renamed variables and coordinates
+
+        Parameters
+        ----------
+        name_dict : dict-like, optional
+            Dictionary whose keys are current variable or coordinate names and
+            whose values are the desired names.
+        inplace : bool, optional
```
Comment (max-sixty) [inline xarray/core/dataset.py]: Given inplace has been deprecated, I don't think we should add it here

File: xarray/core/dataset.py [hunk 4]
```diff
@@ -2285,10 +2287,88 @@ def rename(self, name_dict=None, inplace=None, **names):
...
+                raise ValueError("cannot rename %r because it is not a "
+                                 "dimension in this dataset" % k)
+
+        dims = self._rename_dims(name_dict)
+        indexes = self._rename_indexes(name_dict)
+        return self._replace(dims=dims, indexes=indexes, inplace=inplace)
+
+
+    def rename_vars(self, name_dict=None, inplace=None, **names):
+        """Returns a new object with renamed variables and coordinates
```
Comment (max-sixty) [inline xarray/core/dataset.py]: Maybe technically this should say "variables including coordinates", but this isn't the only place where it's written like this (and let's not slow down this PR with these)

**Comment by max-sixty** [comment]
This is great! Thanks @jukent ! Ping back here if there's anything we can do to help

**Comment by max-sixty** [comment]
And tbc, we should add tests here. That will help clarify the issues above re exactly which items on the dataset each method should rename vs leave.

File: xarray/core/dataset.py (hunk 4 above)
Comment (jukent) [reply xarray/core/dataset.py]: Done

File: xarray/core/dataset.py (hunk 3 above)
Comment (jukent) [reply xarray/core/dataset.py]: Makes sense. I removed it, but it is still floating around in the other functions in the script (swap_dims for example)

**Comment by jukent** [comment]
> And tbc, we should add tests here. That will help clarify the issues above re exactly which items on the dataset each method should rename vs leave.

Yes I need to add tests, but I don't have much experience with this yet. I will ask my team here to help me.

**Comment by jukent** [comment]
> > #3026
> > There is some unexpected behavior. The new rename_dims function does not change the dimensions of the variables. Working on this now.
> 
> Similarly rename_vars will rename the variable dimensions and variables, but not the dataset dimensions. I thought I was dealing with this with the _rename_indexes. But will have to keep working.

This works now! But to do so I created three different renaming variable helper functions (var name only, var and dims, var dims only). I will clean this up since the majority of these three functions is the same.

**Comment by max-sixty** [comment]
@jukent check out the existing rename tests here (they're a bit over-complicated at the beginning): https://github.com/pydata/xarray/blob/master/xarray/tests/test_dataset.py#L2046-L2130

You could do something as simple as this: https://github.com/pydata/xarray/blob/master/xarray/tests/test_dataset.py#L2114-L2119 - i.e. start with a dataset, call a rename method, and compare the dataset with a manually constructed version of the expected result.

Ping back with any issues

**Comment by jukent** [comment]
> @jukent check out the existing rename tests here (they're a bit over-complicated at the beginning): https://github.com/pydata/xarray/blob/master/xarray/tests/test_dataset.py#L2046-L2130
> 
> You could do something as simple as this: https://github.com/pydata/xarray/blob/master/xarray/tests/test_dataset.py#L2114-L2119 - i.e. start with a dataset, call a rename method, and compare the dataset with a manually constructed version of the expected result.
> 
> Ping back with any issues

Thanks @max-sixty. I added some very simple tests. I might copy more of the test_rename functions (if I tried to rename a variable to the same name for example)

**Comment by max-sixty** [comment]
Those look great!

**Comment by max-sixty** [comment]
@jukent I don't think that's our bot

**Pull request closed without merging**